index.d.ts

```diff
@@ -1,47 +1,49 @@
 // Project: Immutability helper
-// TypeScript Version: 2.2
+// TypeScript Version: 2.8
 
 export default update
 
-declare function update<T>(
-  data: ReadonlyArray<T>,
-  query: ArrayOperators<T>,
-): ReadonlyArray<T>
-
-declare function update<T>(
-  data: ReadonlySet<T>,
-  query: SetOperators<T>,
-): ReadonlySet<T>
-
-declare function update<K, V>(
-  data: ReadonlyMap<K, V>,
-  query: MapOperators<K, V>,
-): ReadonlyMap<K, V>
-
-declare function update<T>(data: T, query: Query<T>): T
-
-type Tree<T> = {[K in keyof T]?: Query<T[K]>}
-export type Query<T> =
-  | Tree<T>
-  | ObjectOperators<T>
-  | ArrayOperators<any>
-  | SetOperators<any>
-
-type ObjectOperators<T> =
-  | {$set: any}
-  | {$toggle: Array<keyof T | number>}
-  | {$unset: Array<keyof T | number>}
-  | {$merge: Partial<T>}
-  | {$apply: (old: T) => T}
-  | ((old: T) => any)
-type ArrayOperators<T> =
-  | {$push: T}
-  | {$unshift: T}
-  | {$splice: Array<[number, number]>}
-  | {[customCommand: string]: any}
-
-type MapOperators<K, V> = {$add: Array<[K, V]>} | {$remove: K[]}
-type SetOperators<T> = {$add: T[]} | {$remove: T[]}
+declare function update<T>(target: T, spec: Spec<T>): T
+
+export type Spec<T> =
+  T extends (Array<infer U> | ReadonlyArray<infer U>) ? ArraySpec<U> :
+  T extends (Map<infer K, infer V> | ReadonlyMap<infer K, infer V>) ? MapSpec<K, V> :
+  T extends (Set<infer U> | ReadonlySet<infer U>) ? SetSpec<U> :
+  T extends object ? ObjectSpec<T> :
+  BaseSpec<T>
+
+type BaseSpec<T> =
+  | { $set: T }
+  | { $apply: (v: T) => T }
+  | ((v: T) => T)
+
+type ArraySpec<T> =
+  | BaseSpec<Array<T>>
+  | { $push: T }
+  | { $unshift: T }
+  // Technically splice is variadic, but we stop after the third shape because that usage is very rare and we can't do variadic types here.
+  | { $splice: Array<[number] | [number, number] | [number, number, T]> }
+  // TODO: Wanted this to work...
+  // | { [index: number]: Spec<T> }
```

Comment:
This doesn't work because [type aliases can't refer to themselves except as properties](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-aliases). I don't think there's a way around that.